<a href="https://colab.research.google.com/github/dmorton714/code-You_DA_demos/blob/main/m3w1_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import sqlite3
from sqlite3 import connect

# Data Being Made

****This code is less important. It just generates data for us to use below.****


This code generates a synthetic dataset with 60 entries, each representing a person with randomly assigned attributes such as name, age, gender, race, and income, utilizing NumPy and Pandas for data manipulation.

In [ ]:
names = np.array([f'Person{i}' for i in range(60)])
ages = np.random.randint(18, 65, 60)
genders = np.random.choice(['Male', 'Female'], 60)
races = np.random.choice(['White', 'Black', 'Hispanic', 'Asian', 'Other'], 60)
incomes = np.random.randint(10000, 100000, 60)
df = pd.DataFrame({
    'name': names,
    'age': ages,
    'gender': genders,
    'race': races,
    'income': incomes
})

This code maps real names to corresponding code names in a DataFrame. The provided list 'real_names' is associated with previously coded names, creating a new DataFrame 'code_names' with columns 'real_name' and 'code_name' for reference and comparison in data analysis.

In [ ]:
real_names = [
    "Frank Smith", "Tony Adams", "Amanda Appleseed", "Robert Johnson", "Julia White",
    "Michael Harris", "Sarah Baker", "William Foster", "Emily Turner", "James Carter",
    "Linda Parker", "Thomas Reed", "Jessica Hayes", "Richard Brown", "Olivia Murphy",
    "Charles Anderson", "Sophia Taylor", "David Martin", "Emma Nelson", "Daniel Wright",
    "Ashley Hall", "Matthew Bennett", "Jennifer Scott", "Brian Coleman", "Lauren Evans",
    "Christopher Mitchell", "Grace Stewart", "Joseph Lopez", "Lily Garcia", "Andrew Turner",
    "Natalie Brooks", "Kenneth Young", "Zoey Sanchez", "Samuel King", "Madison Collins",
    "Benjamin Butler", "Chloe Foster", "John Reed", "Victoria Moore", "Nicholas Perry",
    "Ava Simmons", "Ryan Smith", "Hannah Watson", "George Wilson", "Elizabeth Miller",
    "William Hayes", "Penelope Hall", "Jack Cooper", "Abigail Price", "Patrick Green",
    "Mia Russell", "Ethan Turner", "Sophia White", "Kevin Cook", "Zoey Hill",
    "Daniel Turner", "Scarlett Simmons", "Ryan Adams", "Grace Martinez", "Samuel Howard"
]

# Create a DataFrame
code_names = np.array([f'Person{i}' for i in range(60)])

code_names = pd.DataFrame({'real_name': real_names, 'code_name': code_names})

# SQL Connect

The code establishes an in-memory SQLite database using `connect(':memory:')`. It then stores the DataFrames df and code_names as tables named "df" and "code_names" in the SQLite database, enabling SQL-like querying and analysis on the in-memory data. This is particularly useful for temporary data storage and retrieval within the scope of the current session.



In [ ]:
conn = connect(':memory:')
# create tables
df.to_sql("df", conn)
code_names.to_sql("code_names", conn)

60


The sql function provided is a convenience function for executing `SQL` queries on the SQLite database `(conn)`. It takes an SQL query string as input and uses `pd.read_sql()` from the Pandas library to execute the query on the database, returning the results as a Pandas DataFrame. This abstraction simplifies the process of querying the database and obtaining the results, making it more concise and readable in the code.

In [ ]:
# convenient function
def sql(a_string):
    return(pd.read_sql(a_string, conn))

## Table schemas

![Table Schema image](images/schema.jpg)

Make your own diagram: https://app.diagrams.net/

This query retrieves the first 5 rows of all columns from the table named "df" in the SQLite database (conn). It provides a preview of the data in the "df" table, allowing you to examine a small subset of the records to understand the structure and content of the DataFrame stored in the database.

In [ ]:
sql("""SELECT * FROM df LIMIT 5""")

index     name  age  gender      race  income
0      0  Person0   62    Male     Black   20794
1      1  Person1   60    Male     Asian   36256
2      2  Person2   64  Female     Black   79048
3      3  Person3   53  Female  Hispanic   45324
4      4  Person4   33    Male     Black   29210

In [ ]:
sql("""SELECT * FROM code_names LIMIT 5""")

index         real_name code_name
0      0       Frank Smith   Person0
1      1        Tony Adams   Person1
2      2  Amanda Appleseed   Person2
3      3    Robert Johnson   Person3
4      4       Julia White   Person4

This query counts the number of rows in the "df" table and returns a single value representing the total row count. It provides a quick way to retrieve the size of the DataFrame stored in the SQLite database.

In [ ]:
sql('SELECT count(*) FROM df;')

count(*)
0        60

This query is a pragma statement in SQLite that retrieves information about the columns (fields) of the "df" table. It returns a result set with details such as the column name, data type, whether the column can contain NULL values, and the default value, providing insights into the structure of the DataFrame stored in the SQLite database.

In [ ]:
sql('PRAGMA table_info(df);')

cid    name     type  notnull dflt_value  pk
0    0   index  INTEGER        0       None   0
1    1    name     TEXT        0       None   0
2    2     age  INTEGER        0       None   0
3    3  gender     TEXT        0       None   0
4    4    race     TEXT        0       None   0
5    5  income  INTEGER        0       None   0

This query retrieves distinct values from the 'age' column of the "df" table, limiting the result to the first 5 unique values. It provides a sample of unique age values present in the DataFrame, allowing you to examine a subset of distinct ages within the dataset.

In [ ]:
sql("SELECT DISTINCT age FROM df LIMIT 5")

age
0   62
1   60
2   64
3   53
4   33

This query retrieves all columns (*) from the "df" table where the 'age' column values are greater than 60. It filters the data to include only rows where the age exceeds 60, providing a subset of the DataFrame with individuals who are older than 60 years.


In [ ]:
sql("SELECT * FROM df WHERE age > 60")

index      name  age  gender   race  income
0      0   Person0   62    Male  Black   20794
1      2   Person2   64  Female  Black   79048
2      6   Person6   63  Female  Other   38247
3     22  Person22   64  Female  Asian   16527
4     29  Person29   61    Male  Asian   77427
5     34  Person34   64  Female  White   26554
6     39  Person39   63  Female  Black   48268
7     45  Person45   62    Male  Black   93311
8     50  Person50   62  Female  Black   35182
9     55  Person55   64    Male  Black   24735

This query retrieves all columns (*) from the "df" table where both conditions are met: the 'age' column values are greater than 60, and the 'income' column values are greater than 60000. It filters the data to include rows where individuals are older than 60 and have an income exceeding $60,000, providing a subset of the DataFrame that satisfies both criteria.

In [ ]:
sql('SELECT * FROM df WHERE age > 60 AND income > 60000')

index      name  age  gender   race  income
0      2   Person2   64  Female  Black   79048
1     29  Person29   61    Male  Asian   77427
2     45  Person45   62    Male  Black   93311

This query retrieves all columns (*) from the "df" table where the conditions are met: the 'gender' column values are 'Male', the 'race' column values are 'Black', and the 'income' column values are greater than 40000. It filters the data to include rows where individuals are Male, belong to the Black race, and have an income exceeding $40,000, providing a subset of the DataFrame that satisfies all specified criteria.

In [ ]:
sql("SELECT * FROM df WHERE gender = 'Male' AND race = 'Black' AND income > 40000")

index      name  age gender   race  income
0     17  Person17   43   Male  Black   51408
1     23  Person23   32   Male  Black   41459
2     41  Person41   42   Male  Black   68504
3     45  Person45   62   Male  Black   93311

# Joins

## df Table schema

---
| index | name | age | gender | race | income |
|------|------|------|------|------|------|
| 0	| Person0 |	62	| Male	| Black	| 20794 |
| 1 |Person1 | 60 | Male | Asian | 36256 |
| 2 | Person2 | 64 | Female | Black | 79048 |
| 3 | Person3	| 53 | Female | Hispanic | 45324 |
| 4 | Person4	| 33 | Male | Black | 29210 |



## code_names Table Schema


---



| index | real_name | code_name |
|-------|-----------|-----------|
|	0 | Frank Smith | Person0 |
| 1 | Tony Adams | Person1 |
| 2 | Amanda Appleseed | Person2 |
| 3 | Robert Johnson | Person3 |
| 4 | Julia White | Person4 |

This selects all columns from the "df" table (df.*) and the 'real_name' column from the "code_names" table. The JOIN is performed based on the equality of the 'name' column in "df" and the 'code_name' column in "code_names." The result is limited to the first 5 rows. This query combines information from both tables where the names match, providing a unified view of the data that includes the 'real_name' column from the "code_names" table.







In [ ]:
sql("""
SELECT df.*, code_names.real_name
FROM df
JOIN code_names
ON df.name = code_names.code_name
LIMIT 5
""")

index     name  age  gender      race  income         real_name
0      0  Person0   62    Male     Black   20794       Frank Smith
1      1  Person1   60    Male     Asian   36256        Tony Adams
2      2  Person2   64  Female     Black   79048  Amanda Appleseed
3      3  Person3   53  Female  Hispanic   45324    Robert Johnson
4      4  Person4   33    Male     Black   29210       Julia White

The SQL query selects specific columns ('code_name', 'real_name', 'age', 'gender', 'income', 'race') from the combined result of the "df" and "code_names" tables. The JOIN operation is performed based on the equality of the 'name' column in the "df" table and the 'code_name' column in the "code_names" table. The result is limited to the first 5 rows, providing a subset of the combined data with the specified columns from both tables. This query combines information from both tables where the names match, creating a unified view with selected columns.

In [ ]:
sql("""
SELECT code_name, real_name, age, gender, income, race
FROM df JOIN code_names ON df.name = code_names.code_name
LIMIT 5
""")

code_name         real_name  age  gender  income      race
0   Person0       Frank Smith   62    Male   20794     Black
1   Person1        Tony Adams   60    Male   36256     Asian
2   Person2  Amanda Appleseed   64  Female   79048     Black
3   Person3    Robert Johnson   53  Female   45324  Hispanic
4   Person4       Julia White   33    Male   29210     Black

This SQL query retrieves specific columns ('code_name', 'real_name', 'age', 'gender', 'income', 'race') from the combined result of the "df" and "code_names" tables. The JOIN operation is based on the equality of the 'name' column in "df" and the 'code_name' column in "code_names." The query includes conditions to filter the data: age less than 60, income less than $50,000, gender is 'Female', and race is 'Asian.' The results are sorted in descending order by income, and the query returns the top 5 rows meeting these criteria.

In [ ]:
sql("""
SELECT code_name, real_name, age, gender, income, race
FROM df
JOIN code_names
ON df.name = code_names.code_name
WHERE age < 60
AND income < 50000
AND gender = 'Female'
AND race = 'Asian'
ORDER BY income DESC
LIMIT 5
""")


code_name     real_name  age  gender  income   race
0   Person8  Emily Turner   46  Female   45404  Asian
1  Person57    Ryan Adams   31  Female   19837  Asian
2  Person40   Ava Simmons   19  Female   11609  Asian